api/include/opentelemetry/nostd/variant.h

```diff
@@ -0,0 +1,1279 @@
+  struct swapper
+  {
+    template <typename ThisAlt, typename ThatAlt>
+    inline void operator()(ThisAlt &this_alt, ThatAlt &that_alt) const
+    {
+      using std::swap;
+      swap(this_alt.value, that_alt.value);
+    }
+  };
+
+  inline constexpr bool move_nothrow() const
+  {
+    return this->valueless_by_exception() ||
+           std::array<bool, sizeof...(Ts)>{
+               {std::is_nothrow_move_constructible<Ts>::value...}}[this->index()];
+  }
+};
+
+template <typename From, typename To>
+struct is_non_narrowing_convertible
+{
+  template <typename T>
+  static std::true_type test(T(&&)[1]);
+
+  template <typename T>
+  static auto impl(int) -> decltype(test<T>({std::declval<From>()}));
+
+  template <typename>
+  static auto impl(...) -> std::false_type;
+
+  static constexpr bool value = decltype(impl<To>(0))::value;
+};
+
+template <typename Arg,
+          std::size_t I,
+          typename T,
+          bool     = std::is_arithmetic<T>::value,
+          typename = void>
+struct overload_leaf
+{};
+
+template <typename Arg, std::size_t I, typename T>
+struct overload_leaf<Arg, I, T, false>
+{
+  using impl = size_constant<I> (*)(T);
+  operator impl() const { return nullptr; };
+};
+
+template <typename Arg, std::size_t I, typename T>
+struct overload_leaf<Arg,
+                     I,
+                     T,
+                     true
+#if defined(__clang__) || !defined(__GNUC__) || __GNUC__ >= 5
+                     ,
+                     enable_if_t<std::is_same<remove_cvref_t<T>, bool>::value
+                                     ? std::is_same<remove_cvref_t<Arg>, bool>::value
+                                     : is_non_narrowing_convertible<Arg, T>::value>
+#endif
+                     >
+{
+  using impl = size_constant<I> (*)(T);
+  operator impl() const { return nullptr; };
+};
+
+template <typename Arg, typename... Ts>
+struct overload_impl
+{
+private:
+  template <typename>
+  struct impl;
+
+  template <std::size_t... Is>
+  struct impl<index_sequence<Is...>> : overload_leaf<Arg, Is, Ts>...
+  {};
+
+public:
+  using type = impl<index_sequence_for<Ts...>>;
+};
+
+template <typename Arg, typename... Ts>
+using overload = typename overload_impl<Arg, Ts...>::type;
+
+template <typename Arg, typename... Ts>
+using best_match = invoke_result_t<overload<Arg, Ts...>, Arg>;
+
+template <typename T>
+struct is_in_place_index : std::false_type
+{};
+
+template <std::size_t I>
+struct is_in_place_index<in_place_index_t<I>> : std::true_type
+{};
+
+template <typename T>
+struct is_in_place_type : std::false_type
+{};
+
+template <typename T>
+struct is_in_place_type<in_place_type_t<T>> : std::true_type
+{};
+}  // namespace detail
+
+template <typename... Ts>
+class variant
+{
+  static_assert(0 < sizeof...(Ts), "variant must consist of at least one alternative.");
+
+  static_assert(detail::all<!std::is_array<Ts>::value...>::value,
+                "variant can not have an array type as an alternative.");
```

Comment:
[According to the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes), attribute values can be homogeneous arrays. I'm not sure how to best handle this with the variant approach.

Also, I expect attribute array values to be a problem for a bunch of vendors, but that's a different story.

Reply:
Arrays of primitives could be supported with something like
```
using AttributeValue = nostd::variant<
  int,
  double,
  nostd::string_view,
  nostd::span<int>,
  nostd::span<double>,
  nostd::span<nostd::string_view>
>;
```

Reply:
I added support for arrays of primitives. There's an example [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/60/files#diff-86a7cce9832b7f4df9d125b035da4fe8R26-R27)